src/hotspot/cpu/aarch64/c1_MacroAssembler_aarch64.cpp

```diff
@@ -304,7 +304,12 @@ void C1_MacroAssembler::allocate_array(Register obj, Register len, Register t1,
     return;
   }
 
-  membar(StoreStore);
+  // Only add membar if zeroing the array.
+  // If not zeroing, subsequent instructions should populate the array (e.g. copy contents),
+  // and the membar should be set after the array has been populated accordingly.
```

Comment:
Strengthen the language:
```suggestion
  // Only add membar if zeroing the array.
  // If not zeroing, subsequent instructions must fully populate the array (e.g. copy contents),
  // and a membar must be emitted after the array has been populated accordingly. See
  // GraphBuilder::append_alloc_array_copy.
```